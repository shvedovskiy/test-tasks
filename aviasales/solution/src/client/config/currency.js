export const RUSSIAN_ROUBLE = 'RUSSIAN_ROUBLE';
export const US_DOLLAR = 'US_DOLLAR';
export const EURO = 'EURO';

export const currencySymbols = {
  [RUSSIAN_ROUBLE]: '₽',
  [US_DOLLAR]: '$',
  [EURO]: '€',
};

export const currencyAliases = {
  [RUSSIAN_ROUBLE]: 'RUB',
  [US_DOLLAR]: 'USD',
  [EURO]: 'EUR',
};
