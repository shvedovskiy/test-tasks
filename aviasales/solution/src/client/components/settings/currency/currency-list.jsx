import React from 'react';

import { currencyAliases } from '~/config/currencies'; // eslint-disable-line import/no-unresolved, import/extensions


const CurrencyList = ({ currency, handleChangeCurrency }) => {
  function handleItemChange(e) {
    const newCurrency = e.target.value;
    handleChangeCurrency(newCurrency);
  }
  return (
    <div>
      {
        Object.keys(currencyAliases).map(currencyKey => (
          <label key={currencyKey}>
            {currencyAliases[currencyKey]}
            <input
              type="radio"
              value={currencyKey}
              checked={currencyKey === currency}
              onChange={handleItemChange}
            />
          </label>
        ))
      }
    </div>
  );
};

export default CurrencyList;
