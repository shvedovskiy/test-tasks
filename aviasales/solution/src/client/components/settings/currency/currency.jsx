import React, { Component } from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';

/* eslint-disable import/first, import/no-unresolved, import/extensions */
import CurrencyList from './currency-list';
import currencyService from '~/services/currency';
import { RUSSIAN_ROUBLE, currencyAliases } from '~/config/currency';
import { getCurrency } from '~/store/selectors';
import { changeCurrency } from '~/store/settings/actions';
/* eslint-enable import/first, import/no-unresolved, import/extensions */


class Currency extends Component {
  state = {
    isFetched: false,
  };

  componentDidMount() {
    currencyService.fetchCurrencies()
      .then(() => {
        this.setState({ isFetched: true });
      });
  }

  render() {
    let { currency, handleChangeCurrency } = this.props;
    const ratesNames = currencyService.getRatesNames();
    const aliases = _.pickBy(currencyAliases, value => ratesNames.includes(value));

    if (!aliases[currency]) {
      currency = RUSSIAN_ROUBLE;
    }

    return (
      <CurrencyList
        aliases={aliases}
        currency={currency}
        handleChangeCurrency={handleChangeCurrency}
      />
    );
  }
}

const mapStateToProps = state => ({
  currency: getCurrency(state),
});

export default connect(mapStateToProps, { handleChangeCurrency: changeCurrency })(Currency);
