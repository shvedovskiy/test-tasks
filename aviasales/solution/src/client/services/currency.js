// @flow
import _ from 'lodash';

import { RUSSIAN_ROUBLE, currencyAliases } from 'src/config/currency';
import { FIXER_API_KEY } from 'shared/config';


type RatesType = {
  [string]: number,
};

class CurrencyService {
  API_ENDPOINT = 'http://data.fixer.io/api';

  symbols = Object.values(currencyAliases).join(',');

  rates: RatesType = {
    [currencyAliases[RUSSIAN_ROUBLE]]: 1,
  };

  async fetchCurrencies() {
    const url = `${this.API_ENDPOINT}/latest?access_key=${FIXER_API_KEY}&symbols=${this.symbols}`;
    let response;
    let fetchData: { rates: ?RatesType } = { rates: null };

    try {
      response = await fetch(url, { method: 'GET' });
      if (response.ok) {
        fetchData = await response.json();
      }
    } catch (err) {
      console.error('Currencies fetching failed'); // eslint-disable-line no-console
    }

    const { rates } = fetchData;
    if (rates !== null && rates !== undefined) {
      const coefficient = rates[currencyAliases[RUSSIAN_ROUBLE]];
      const newRates = _.mapValues(rates, value => Number((value / coefficient).toFixed(2)));
      this.rates = {
        ...this.rates,
        ...newRates,
      };
    }
  }

  getPrice(price: string, currency: string): ?number {
    const rates = this.getRates();
    const convertIndex: ?number = rates[currencyAliases[currency]];
    let result: ?number;

    if (convertIndex !== null && convertIndex !== undefined) {
      result = Number.parseFloat(price) * convertIndex;
    }
    return (result && !Number.isNaN(result)) ? result : null;
  }

  getRates(): RatesType {
    return this.rates;
  }

  getRatesNames(): Array<string> {
    return Object.keys(this.rates);
  }
}

export default new CurrencyService();
