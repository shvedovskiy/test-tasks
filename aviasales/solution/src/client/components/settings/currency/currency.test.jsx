/* eslint-disable no-undef */
import React from 'react';

import { Currency } from './currency';


describe('<Currency />', () => {
  const props = {
    selectedCurrency: 'Russian Rouble',
    handleChangeCurrency: jest.fn(),
  };
  const currency = shallow(<Currency {...props} />);

  it('renders properly', () => {
    expect(currency).toMatchSnapshot();
  });
});
