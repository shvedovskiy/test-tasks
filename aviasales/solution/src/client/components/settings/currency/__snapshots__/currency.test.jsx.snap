// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Currency /> renders properly 1`] = `
<CurrencyList
  aliases={
    Object {
      "Russian Rouble": "RUB",
    }
  }
  handleChangeCurrency={[MockFunction]}
  selectedCurrency="Russian Rouble"
/>
`;
